### DISCUSSION TITLE
pymagnum mqtt home assistant

### DISCUSSION THREAD

**Discussion by garylankford8** [CharlesGodwin/pymagnum#65] [Q&A]
I have a debian server installed on an orange pi 3 lts and can ssh to it from Termius on Android or Putty on linux.

I have tested the pi/MS4448pae connection, and /dev/ttyUSB0 is active and magtest and magdump run to completion.

I have copied the examples to the pi and 'sudo python3 magsample.py' runs. The example that I think I need to publish mqtt data for the mosquito broker on Home Assistant I think is mqttlogger-2.py.

I have installed mosquito and mosquitto-clients on the pi and have run 

sudo systemctl enable mosquitto.service
sudo systemctl start mosquitto.service 
sudo systemctl status mosquitto.service

and it is Active.

When I run 'sudo python3 mqttlogger-2.py', I get

/home/gary/mqttlogger-2.py:133: DeprecationWarning: Callback API version 1 is deprecated,
update to latest version
  client = mqtt.Client(client_id=uuidstr, clean_session=False)

I'm an old guy... I got this far... any help? I'm away from the server right now, but having typed this already, I'm gonna post it and clarify as necessary later... Sorry, and thank you in advance!

**Comment by CharlesGodwin** [comment]
Thanks for your interest in my software.
don't beat yourself up for being "an old guy". I too am an old guy.

I'm going to investigate the warning message you got. **Note** this is  warning, not an error. The code should work.

Can you tell me what versions of software you have? run these two commands

`uname -a`
`python3 --version`

You also write that you prefix the commands with `sudo`. I don't normally do that, is it necessary on you machine. I do all my testing on Raspberry Pi or Windows and have no experience with the version of Debian on your Orange Pi.

Stay tuned

**Comment by CharlesGodwin** [comment]
Try this version. I changed 2 lines. The file is compressed to attach it.
To use it

1. download file to directory you want
2. uncompress with this command
`gunzip mqttlogger-2.py.gz`

If it works, I'll revise the documentation and publish a new version.
[mqttlogger-2.py.gz](https://github.com/user-attachments/files/16377289/mqttlogger-2.py.gz)

**Comment by CharlesGodwin** [comment]
I tested with a test program as I'm not near my Pi system and I do get the warning, with or without my revisions, but the code still works.

**Comment by CharlesGodwin** [comment]
I added a [warning ](https://github.com/CharlesGodwin/pymagnum/blob/master/examples/README.md)to the examples `README.md`

**Reply by garylankford8** [reply]
The whole OS installation on the Pi was weird to me and left me with only root access. Once I gave up on trying to use /dev/ttyS0 and used a usb adapter, python found /dev/ttyUSB0 easy enough and magtest ran. Not liking running root, I added a user account, un-installed pymagnum as root and reinstalled as user, but then magtest still wouldn't run without sudo... These are all problems with the orangepi and the OS which I will try to rectify.

$ uname -a
Linux orangepi3-lts 5.16.17-sun50iw6 #3.0.8 SMP Tue Sep 6 19:19:38 CST 2022 aarch64 GNU/Linux

$ python3 --version
Python 3.9.2

This issue I am seeing on the pi has something to do with the version of paho.python.

The new mqttlogger-2.py version didn't fix it.

I installed pymagnum on my Linux Mint laptop (python 3.10.12),  connected the USB-RS485 cable to my MS4448pae and ran the previous version of mqttlogger.py without sudo and it runs without error saying...

"Publishing to broker:localhost:1883 Every:60 seconds. Using ['devadded a user account/ttyuSB0']"

This, despite my having changed pymagnum.opt broker to core-mosquitto:1883 which is my Home Assistant broker and HA is not seeing the published data from the laptop.

I barely understand what I am doing here...

Maybe there is some mosquitto setup issue on my HA, but it is receiving data published by my Solar Assistant Pi from my SOK batteries...

**Reply by CharlesGodwin** [reply]
I tested and read the documents. All it is, is a warning message. When I tested I got the warning but the code worked. The software documents say it will work in this version.

I'm surprised about sudo but if it works for you, go with it

**Reply by CharlesGodwin** [reply]
I am working on a message free version but it won't be ready before Monday

**Reply by CharlesGodwin** [reply]
Do you have the broker setup properly. The default requires user credentials. I am not an expert on setting up MQTT brokers

**Reply by CharlesGodwin** [reply]
Try running mqttlogger-2.py with the --trace option

**Reply by garylankford8** [reply]
Message 1 published
Message 2 published

...1 min
Message 3 published
Message 4 published

...so on

Where should the download examples be placed? Should the opt be in /opt?

**Reply by CharlesGodwin** [reply]
Try running mqttlogger-2.py with the --trace option I'll look at this Saturday

**Comment by CharlesGodwin** [comment]
Let's test where the problem lies. 

Before you try sending data to HA lets see if it's working by sending messages to the broker on the Pi

- start a ssh session on the Pi and enter this command (you need user credentials if that's how you are set up, otherwise the command will ignore the credentials). 
`mosquitto_sub -h localhost -t "#" -v -u user -P password`
 Do not close this session as it will display all output from  mqttlogger-2.
- Start another ssh session
- Change to the folder where you put mqttlogger-2.py
- Start this program with the defalult values using local host
 `sudo python3 mqttlogger-2.py --username user --password password --device /dev/ttyUSB0`
- Switch to other ssh session and you should see a data stream similar to this
 `magnum/inverter {"datetime":"2024-07-26T07:52:39-04:00","device":"INVERTER","data":{"revision":"5.1","mode":64,"mode_text":"INVERT","fault":0,"fault_text":"None","vdc":24.6,"adc":6.0,"VACout":120.0,"VACin":0.0,"invled":1,"invled_text":"On","chgled":0,"chgled_text":"Off","bat":17.0,"tfmr":37.0,"fet":31.0,"model":107,"model_text":"MS4024PAE","stackmode":0,"stackmode_text":"Stand Alone","AACin":0.0,"AACout":2.0,"Hz":60.0}}`
- Let me know what happens. Please be sure to send the output from mqttlogger-2 that starts with `Options:`. Delete password if you wish but send the rest.
- If this works the problem is with the data going to the broker on HA.

**Comment by CharlesGodwin** [comment]
you asked. 
_Where should the download examples be placed? Should the opt be in /opt?_

There is no required location for these files. Place them anywhere that's easy for you to use.

**Comment by CharlesGodwin** [comment]
`mqttlogger-2.py` has been revised to remove warning message and will only work with up to date paho-mqtt package

**Reply by garylankford8** [reply]
I updated my pi to Debian Bookworm which includes Python3.11. The message is naturally gone using the new logger, but there is an unterminated literal at I think line 36 that I fixed. 

The logger is publishing and the home assistant mosquitto broker addon sees it, but I 'm still not getting data in the mqtt integration.

Getting closer though! Thanks

**Comment by CharlesGodwin** [comment]
I'm glad my software is working, but I have reviewed and re-tested the program and I do not have a problem with an unterminated literal at line 36 or anywhere else. Can you provide details.
The version in github has this at line 36 and 37, which is a legitimate statement
<pre>
        raise Exception(f"Connection failed. {
                        ReasonCode.getName(reason_code)}")
</pre>

**Reply by garylankford8** [reply]
When I run it, it stops at that line with the error. Apparently, my version of Python doesn't like the split line without it being triple quoted. All I did was backspace the 2nd line to solve the problem.

raise Exception(f"Connection failed. {ReasonCode.getName(reason_code)}")

**Comment by CharlesGodwin** [comment]
fascinating. thanks

**Reply by garylankford8** [reply]
Something about multiple line code being interpreted as an unterminated literal, even though the code was terminated properly in itself...

Really, all this stuff is at my limit!

**Reply by CharlesGodwin** [reply]
This is not your fault. It seems to be a bug/feature in early versions of Python 3. Python 3.12 can handle it. I'll update the source later today.

**Comment by CharlesGodwin** [comment]
Republished mqttlogger-2.py to comply with earlier python 3 version.

@garylankford8  Can I close this?

**Reply by garylankford8** [reply]
Yes
